diff --git a/src/WebSocket.js b/src/WebSocket.js
index c12db834fd4a30da0c46b3b623e961dcc8003f7b..6c0f617d875d175e69c9160a4a870933e483921c 100644
--- a/src/WebSocket.js
+++ b/src/WebSocket.js
@@ -3,7 +3,7 @@
 let erlpack;

 try {
-  erlpack = require('erlpack');
+  erlpack = require('@yukikaze-bot/erlpack');
   if (!erlpack.pack) erlpack = null;
 } catch {} // eslint-disable-line no-empty

diff --git a/src/client/Client.js b/src/client/Client.js
index 3079c3b34f2682389c63a5dbef4446bc6b2781cc..aa520f9a6b630846d4dace3818693b2fe24024ca 100644
--- a/src/client/Client.js
+++ b/src/client/Client.js
@@ -10,7 +10,7 @@ const ChannelManager = require('../managers/ChannelManager');
 const GuildManager = require('../managers/GuildManager');
 const UserManager = require('../managers/UserManager');
 const ShardClientUtil = require('../sharding/ShardClientUtil');
-const ClientPresence = require('../structures/ClientPresence');
+const Structures = require('../util/Structures');
 const GuildPreview = require('../structures/GuildPreview');
 const GuildTemplate = require('../structures/GuildTemplate');
 const Invite = require('../structures/Invite');
@@ -119,6 +119,8 @@ class Client extends BaseClient {
      */
     this.channels = new ChannelManager(this);

+    const ClientPresence = Structures.get('ClientPresence');
+
     /**
      * The presence of the Client
      * @private
diff --git a/src/client/actions/InteractionCreate.js b/src/client/actions/InteractionCreate.js
index 85072d059e2bb4fef470b248e23ac1bd2941f37c..141c883013cdbab784a5e1062a54b147111d905b 100644
--- a/src/client/actions/InteractionCreate.js
+++ b/src/client/actions/InteractionCreate.js
@@ -1,9 +1,7 @@
 'use strict';

 const Action = require('./Action');
-const ButtonInteraction = require('../../structures/ButtonInteraction');
-const CommandInteraction = require('../../structures/CommandInteraction');
-const SelectMenuInteraction = require('../../structures/SelectMenuInteraction');
+const Structures = require('../../util/Structures');
 const { Events, InteractionTypes, MessageComponentTypes } = require('../../util/Constants');

 let deprecationEmitted = false;
@@ -18,15 +16,15 @@ class InteractionCreateAction extends Action {
     let InteractionType;
     switch (data.type) {
       case InteractionTypes.APPLICATION_COMMAND:
-        InteractionType = CommandInteraction;
+        InteractionType = Structures.get('CommandInteraction');
         break;
       case InteractionTypes.MESSAGE_COMPONENT:
         switch (data.data.component_type) {
           case MessageComponentTypes.BUTTON:
-            InteractionType = ButtonInteraction;
+            InteractionType = Structures.get('ButtonInteraction');
             break;
           case MessageComponentTypes.SELECT_MENU:
-            InteractionType = SelectMenuInteraction;
+            InteractionType = Structures.get('SelectMenuInteraction');
             break;
           default:
             client.emit(
diff --git a/src/client/actions/VoiceStateUpdate.js b/src/client/actions/VoiceStateUpdate.js
index f750e048b95dd3a1e890625fa79aadceac93731f..8925ba375c54015abb298bd4120c3993edcb4200 100644
--- a/src/client/actions/VoiceStateUpdate.js
+++ b/src/client/actions/VoiceStateUpdate.js
@@ -1,7 +1,7 @@
 'use strict';

 const Action = require('./Action');
-const VoiceState = require('../../structures/VoiceState');
+const Structures = require('../../util/Structures');
 const { Events } = require('../../util/Constants');

 class VoiceStateUpdate extends Action {
@@ -9,6 +9,7 @@ class VoiceStateUpdate extends Action {
     const client = this.client;
     const guild = client.guilds.cache.get(data.guild_id);
     if (guild) {
+      const VoiceState = Structures.get('VoiceState');
       // Update the state
       const oldState =
         guild.voiceStates.cache.get(data.user_id)?._clone() ?? new VoiceState(guild, { user_id: data.user_id });
diff --git a/src/index.js b/src/index.js
index 34cc9f298c68600ca4c4d2afa482c04adae88203..5b287a38b725fe6295d843b217d3e1f91c98a5b7 100644
--- a/src/index.js
+++ b/src/index.js
@@ -109,6 +109,7 @@ module.exports = {
   Role: require('./structures/Role'),
   SelectMenuInteraction: require('./structures/SelectMenuInteraction'),
   Sticker: require('./structures/Sticker'),
+  Structures: require('./util/Structures'),
   StoreChannel: require('./structures/StoreChannel'),
   StageChannel: require('./structures/StageChannel'),
   Team: require('./structures/Team'),
diff --git a/src/managers/DataManager.js b/src/managers/DataManager.js
index 6a5c14ca2015ec0d12b292d6b86673ce569cc2b9..ff0f70674719194a2944aff1a4cd7428446a3307 100644
--- a/src/managers/DataManager.js
+++ b/src/managers/DataManager.js
@@ -3,6 +3,8 @@
 const BaseManager = require('./BaseManager');
 const { Error } = require('../errors');

+let Structures;
+
 /**
  * Manages the API methods of a data model along with a collection of instances.
  * @extends {BaseManager}
@@ -12,6 +14,8 @@ class DataManager extends BaseManager {
   constructor(client, holds) {
     super(client);

+    if (!Structures) Structures = require('../util/Structures');
+
     /**
      * The data structure belonging to this manager.
      * @name DataManager#holds
@@ -19,7 +23,7 @@ class DataManager extends BaseManager {
      * @private
      * @readonly
      */
-    Object.defineProperty(this, 'holds', { value: holds });
+     Object.defineProperty(this, 'holds', { value: Structures.get(holds.name) ?? holds });
   }

   /**
diff --git a/src/structures/Channel.js b/src/structures/Channel.js
index 9d1825048ffecd339b8abcdc7d78a9ed7a3f504d..cc79f3e115cffecf89a09f8f48c02d3026a403e9 100644
--- a/src/structures/Channel.js
+++ b/src/structures/Channel.js
@@ -1,14 +1,6 @@
 'use strict';

 const Base = require('./Base');
-let CategoryChannel;
-let DMChannel;
-let NewsChannel;
-let StageChannel;
-let StoreChannel;
-let TextChannel;
-let ThreadChannel;
-let VoiceChannel;
 const { ChannelTypes, ThreadChannelTypes } = require('../util/Constants');
 const SnowflakeUtil = require('../util/SnowflakeUtil');

@@ -127,18 +119,12 @@ class Channel extends Base {
   }

   static create(client, data, guild, allowUnknownGuild) {
-    if (!CategoryChannel) CategoryChannel = require('./CategoryChannel');
-    if (!DMChannel) DMChannel = require('./DMChannel');
-    if (!NewsChannel) NewsChannel = require('./NewsChannel');
-    if (!StageChannel) StageChannel = require('./StageChannel');
-    if (!StoreChannel) StoreChannel = require('./StoreChannel');
-    if (!TextChannel) TextChannel = require('./TextChannel');
-    if (!ThreadChannel) ThreadChannel = require('./ThreadChannel');
-    if (!VoiceChannel) VoiceChannel = require('./VoiceChannel');
+    const Structures = require('../util/Structures');

     let channel;
     if (!data.guild_id && !guild) {
       if ((data.recipients && data.type !== ChannelTypes.GROUP_DM) || data.type === ChannelTypes.DM) {
+        const DMChannel = Structures.get('DMChannel');
         channel = new DMChannel(client, data);
       } else if (data.type === ChannelTypes.GROUP_DM) {
         const PartialGroupDMChannel = require('./PartialGroupDMChannel');
@@ -150,32 +136,39 @@ class Channel extends Base {
       if (guild || allowUnknownGuild) {
         switch (data.type) {
           case ChannelTypes.GUILD_TEXT: {
+            const TextChannel = Structures.get('TextChannel');
             channel = new TextChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_VOICE: {
+            const VoiceChannel = Structures.get('VoiceChannel');
             channel = new VoiceChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_CATEGORY: {
+            const CategoryChannel = Structures.get('CategoryChannel');
             channel = new CategoryChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_NEWS: {
+            const NewsChannel = Structures.get('NewsChannel');
             channel = new NewsChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_STORE: {
+            const StoreChannel = Structures.get('StoreChannel');
             channel = new StoreChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_STAGE_VOICE: {
+            const StageChannel = Structures.get('StageChannel');
             channel = new StageChannel(guild, data, client);
             break;
           }
           case ChannelTypes.GUILD_NEWS_THREAD:
           case ChannelTypes.GUILD_PUBLIC_THREAD:
           case ChannelTypes.GUILD_PRIVATE_THREAD: {
+            const ThreadChannel = Structures.get('ThreadChannel');
             channel = new ThreadChannel(guild, data, client);
             if (!allowUnknownGuild) channel.parent?.threads.cache.set(channel.id, channel);
             break;
diff --git a/src/structures/ClientUser.js b/src/structures/ClientUser.js
index 9d18ae5e8835d9f50308bb748cc092d986200cc7..717e37daae8a36bd5b740acb190e294afcf4270c 100644
--- a/src/structures/ClientUser.js
+++ b/src/structures/ClientUser.js
@@ -1,13 +1,13 @@
 'use strict';

-const User = require('./User');
+const Structures = require('../util/Structures');
 const DataResolver = require('../util/DataResolver');

 /**
  * Represents the logged in client's Discord user.
  * @extends {User}
  */
-class ClientUser extends User {
+class ClientUser extends Structures.get('User') {
   constructor(client, data) {
     super(client, data);
     this._typing = new Map();
diff --git a/src/structures/GuildMember.js b/src/structures/GuildMember.js
index 8d10126aaa773a8c9ff90dfc00e06408abc97f9b..e3036a3efd8a91596e13111b2e40cb875b7064d4 100644
--- a/src/structures/GuildMember.js
+++ b/src/structures/GuildMember.js
@@ -1,7 +1,7 @@
 'use strict';

 const Base = require('./Base');
-const VoiceState = require('./VoiceState');
+let Structures;
 const TextBasedChannel = require('./interfaces/TextBasedChannel');
 const { Error } = require('../errors');
 const GuildMemberRoleManager = require('../managers/GuildMemberRoleManager');
@@ -109,6 +109,8 @@ class GuildMember extends Base {
    * @readonly
    */
   get voice() {
+    if (!Structures) Structures = require('../util/Structures');
+    const VoiceState = Structures.get('VoiceState');
     return this.guild.voiceStates.cache.get(this.id) ?? new VoiceState(this.guild, { user_id: this.id });
   }

diff --git a/src/util/Structures.js b/src/util/Structures.js
new file mode 100644
index 0000000000000000000000000000000000000000..33343e7e12948d21f0e38cdd224b68d360c47934
--- /dev/null
+++ b/src/util/Structures.js
@@ -0,0 +1,122 @@
+'use strict';
+
+/**
+ * An extendable structure:
+ * * **`GuildEmoji`**
+ * * **`DMChannel`**
+ * * **`TextChannel`**
+ * * **`VoiceChannel`**
+ * * **`CategoryChannel`**
+ * * **`NewsChannel`**
+ * * **`StoreChannel`**
+ * * **`StageChannel`**
+ * * **`ThreadChannel`**
+ * * **`GuildMember`**
+ * * **`ThreadMember`**
+ * * **`Guild`**
+ * * **`Message`**
+ * * **`MessageReaction`**
+ * * **`Presence`**
+ * * **`ClientPresence`**
+ * * **`VoiceState`**
+ * * **`Role`**
+ * * **`User`**
+ * * **`CommandInteraction`**
+ * * **`ButtonInteraction`**
+ * * **`StageInstance`**
+ * * **`SelectMenuInteraction`**
+ * @typedef {string} ExtendableStructure
+ */
+
+/**
+ * Allows for the extension of built-in Discord.js structures that are instantiated by {@link BaseManager Managers}.
+ */
+class Structures extends null {
+  /**
+   * Retrieves a structure class.
+   * @param {string} structure Name of the structure to retrieve
+   * @returns {Function}
+   */
+  static get(structure) {
+    if (typeof structure === 'string') return structures[structure];
+    throw new TypeError(`"structure" argument must be a string (received ${typeof structure})`);
+  }
+
+  /**
+   * Extends a structure.
+   * <warn> Make sure to extend all structures before instantiating your client.
+   * Extending after doing so may not work as expected. </warn>
+   * @param {ExtendableStructure} structure Name of the structure class to extend
+   * @param {Function} extender Function that takes the base class to extend as its only parameter and returns the
+   * extended class/prototype
+   * @returns {Function} Extended class/prototype returned from the extender
+   * @example
+   * const { Structures } = require('discord.js');
+   *
+   * Structures.extend('Guild', Guild => {
+   *   class CoolGuild extends Guild {
+   *     constructor(client, data) {
+   *       super(client, data);
+   *       this.cool = true;
+   *     }
+   *   }
+   *
+   *   return CoolGuild;
+   * });
+   */
+  static extend(structure, extender) {
+    if (!structures[structure]) throw new RangeError(`"${structure}" is not a valid extensible structure.`);
+    if (typeof extender !== 'function') {
+      const received = `(received ${typeof extender})`;
+      throw new TypeError(
+        `"extender" argument must be a function that returns the extended structure class/prototype ${received}.`,
+      );
+    }
+
+    const extended = extender(structures[structure]);
+    if (typeof extended !== 'function') {
+      const received = `(received ${typeof extended})`;
+      throw new TypeError(`The extender function must return the extended structure class/prototype ${received}.`);
+    }
+
+    if (!(extended.prototype instanceof structures[structure])) {
+      const prototype = Object.getPrototypeOf(extended);
+      const received = `${extended.name ?? 'unnamed'}${prototype.name ? ` extends ${prototype.name}` : ''}`;
+      throw new Error(
+        'The class/prototype returned from the extender function must extend the existing structure class/prototype' +
+          ` (received function ${received}; expected extension of ${structures[structure].name}).`,
+      );
+    }
+
+    structures[structure] = extended;
+    return extended;
+  }
+}
+
+const structures = {
+  GuildEmoji: require('../structures/GuildEmoji'),
+  DMChannel: require('../structures/DMChannel'),
+  TextChannel: require('../structures/TextChannel'),
+  VoiceChannel: require('../structures/VoiceChannel'),
+  CategoryChannel: require('../structures/CategoryChannel'),
+  NewsChannel: require('../structures/NewsChannel'),
+  StoreChannel: require('../structures/StoreChannel'),
+  StageChannel: require('../structures/StageChannel'),
+  ThreadChannel: require('../structures/ThreadChannel'),
+  GuildMember: require('../structures/GuildMember'),
+  ThreadMember: require('../structures/ThreadMember'),
+  Guild: require('../structures/Guild'),
+  Message: require('../structures/Message'),
+  MessageReaction: require('../structures/MessageReaction'),
+  Presence: require('../structures/Presence').Presence,
+  ClientPresence: require('../structures/ClientPresence'),
+  VoiceState: require('../structures/VoiceState'),
+  Role: require('../structures/Role'),
+  User: require('../structures/User'),
+  CommandInteraction: require('../structures/CommandInteraction'),
+  ButtonInteraction: require('../structures/ButtonInteraction'),
+  SelectMenuInteraction: require('../structures/SelectMenuInteraction'),
+  StageInstance: require('../structures/StageInstance'),
+};
+
+module.exports = Structures;
diff --git a/typings/index.d.ts b/typings/index.d.ts
index 4f78a786d27a850007d9434bc0879cab65841b3e..bf72ceed1c697877978a68f98ca6fc8f268b81ec 100644
--- a/typings/index.d.ts
+++ b/typings/index.d.ts
@@ -993,6 +993,43 @@ export class Message extends Base {
   public unpin(): Promise<Message>;
 }

+export class Structures extends null {
+  private constructor();
+  public static get<K extends keyof Extendable>(structure: K): Extendable[K];
+  public static get(structure: string): (...args: any[]) => void;
+  public static extend<K extends keyof Extendable, T extends Extendable[K]>(
+    structure: K,
+    extender: (baseClass: Extendable[K]) => T,
+  ): T;
+  public static extend<T extends (...args: any[]) => void>(
+    structure: string,
+    extender: (baseClass: typeof Function) => T,
+  ): T;
+}
+
+interface Extendable {
+  GuildEmoji: typeof GuildEmoji;
+  DMChannel: typeof DMChannel;
+  TextChannel: typeof TextChannel;
+  VoiceChannel: typeof VoiceChannel;
+  CategoryChannel: typeof CategoryChannel;
+  NewsChannel: typeof NewsChannel;
+  StoreChannel: typeof StoreChannel;
+  ThreadChannel: typeof ThreadChannel;
+  GuildMember: typeof GuildMember;
+  ThreadMember: typeof ThreadMember;
+  Guild: typeof Guild;
+  Message: typeof Message;
+  MessageReaction: typeof MessageReaction;
+  Presence: typeof Presence;
+  VoiceState: typeof VoiceState;
+  Role: typeof Role;
+  User: typeof User;
+  CommandInteraction: typeof CommandInteraction;
+  ButtonInteraction: typeof ButtonInteraction;
+  SelectMenuInteraction: typeof SelectMenuInteraction;
+}
+
 export class MessageActionRow extends BaseMessageComponent {
   public constructor(data?: MessageActionRow | MessageActionRowOptions);
   public type: 'ACTION_ROW';
